aten/src/ATen/native/Sorting.cpp

```diff
@@ -951,7 +951,29 @@ TORCH_IMPL_FUNC(sort_stable_out)
     indices.zero_();
   } else {
     dim = maybe_wrap_dim(dim, self.dim());
-    sort_stub(self.device().type(), self, values, indices, dim, descending, stable.value_or(false));
+
+    c10::MaybeOwned<Tensor> indices_tmp;
+    auto dtype_indices = at::ScalarType::Long;
+    const auto sort_size = self.dim() > 0 ? self.size(dim) : 1;
+    if (sort_size - 1 <= std::numeric_limits<uint8_t>::max()) {
+      dtype_indices = at::ScalarType::Byte;
+    } else if (sort_size - 1 <= std::numeric_limits<uint16_t>::max()) {
+      dtype_indices = at::ScalarType::UInt16;
+    } else if (sort_size - 1 <= std::numeric_limits<uint32_t>::max()) {
+      dtype_indices = at::ScalarType::UInt32;
+    }
+
+    if (dtype_indices == at::ScalarType::Long) {
+      indices_tmp = c10::MaybeOwned<Tensor>::borrowed(indices);
+    } else {
+      indices_tmp = c10::MaybeOwned<Tensor>::owned(indices.to(dtype_indices));
+    }
+
+    sort_stub(self.device().type(), self, values, *indices_tmp, dim, descending, stable.value_or(false));
+
+    if (!indices_tmp->is_same(indices)) {
+      indices.copy_(*indices_tmp);
```

Comment:
Any benchmarks to show this copy doesn't massively slow things down? I'm guessing not, but would be good to have some speed improvements in addition to this. Also, how the hell is copy being used a const ref? Doesn't it mutate state?

Reply:
1. Let me have some time to think about performance improvement..

2. const ref tensor.copy_
https://github.com/pytorch/pytorch/blob/2055b66006040424bf6ffeeec8710f1a42244d08/aten/src/ATen/native/Sorting.cpp#L941-L948

    Const ref tensor.copy_ in 948 line is already called in the existing code. I'll think about it too.